### PR TITLE
(LTH-12) Extract facter Windows helpers and wrappers to Leatherman

### PR DESCRIPTION
**Pull request by Magisus**
Move the registry, user, process, system_error, and wmi Windows helper classes from facter to Leatherman, creating a Windows utility library for use in other projects.

### CODE REVIEW

File: windows/inc/leatherman/windows/scoped_resource.hpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
+/**
+ * @file
+ * Declares the base class for scoped resources.
+ */
+#pragma once
+
+#include <functional>
+
+namespace leatherman { namespace windows {
+    /**
```
Comment (kylog) [inline windows/inc/leatherman/windows/scoped_resource.hpp]: Is this windows-specific?

File: windows/inc/leatherman/windows/user.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+     * tasks restricted to Administrators.
+     * @return True if the current process has Administrator privileges, otherwise false.
+     */
+    bool is_admin();
+
+    /**
+     * Query token membership to determine whether the current user is a member of the Administrators group.
+     * @return True if user is an Administrator, otherwise false.
+     */
+    bool check_token_membership();
```
Comment (kylog) [inline windows/inc/leatherman/windows/user.hpp]: Today I only see this used in the `is_admin` implementation. Is it of general utility? If not, should we not expose it? 

Either way, is there a better name for the function? Like `in_admin_group`? Otoh if `check_token_membership` will be intuitively obvious to someone doing windows work, then just ignore me :)

**Comment by Magisus** [comment]
No, in facter it was outside of the windows stuff. And actually, apparently the file that included it (wmi.cc) doesn't even use it. So I guess it can be removed from this library? I left the original in facter/inc/facter/util, because I figured it was used elsewhere.

File: windows/inc/leatherman/windows/wmi.hpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+         * Identifier for the WMI property OtherTypeDescription
+         */
+        constexpr static char const* othertypedescription = "OtherTypeDescription";
+
+        /**
+         * Case-insensitive string comparison.
+         */
+        struct ciless : std::binary_function<std::string, std::string, bool>
```
Comment (kylog) [inline windows/inc/leatherman/windows/wmi.hpp]: Extract from the wmi namespace?

File: windows/inc/leatherman/windows/scoped_resource.hpp (hunk 1 above)
Comment (kylog) [inline windows/inc/leatherman/windows/scoped_resource.hpp]: Yeah, I think `scoped_resource` isn't windows specific, but of broad utility. If so, let's see if we can pull up into a non-windows util header somewhere.

File: windows/inc/leatherman/windows/user.hpp (hunk 2 above)
Comment (MikaelSmith) [inline windows/inc/leatherman/windows/user.hpp]: These were extracted from https://github.com/puppetlabs/puppet/blob/003d2b73b4730953092bb7e8b7bc723d2210c999/lib/puppet/util/windows/user.rb, and will likely be needed in a future native client. Both are marked module functions, so I suspect they would be useful elsewhere.

File: util/inc/leatherman/util/scoped_resource.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
+/**
+ * @file
+ * Declares the base class for scoped resources.
+ */
+#pragma once
+
+#include <functional>
+
+namespace leatherman { namespace windows {
```
Comment (MikaelSmith) [inline util/inc/leatherman/util/scoped_resource.hpp]: This probably shouldn't be in the windows namespace.

File: util/inc/leatherman/util/scoped_resource.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        std::function<void(T&)> _deleter;
+
+     private:
+        void* operator new(size_t) = delete;
+        void operator delete(void*) = delete;
+        void* operator new[](size_t) = delete;
+        void operator delete[](void* ptr) = delete;
+    };
+
+}}  // namespace leatherman::windows
```
Comment (MikaelSmith) [inline util/inc/leatherman/util/scoped_resource.hpp]: Same note about windows here.

File: util/inc/leatherman/util/strings.hpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/**
+ * @file
+ * Declares utility functions for dealing with strings.
+ */
+#pragma once
+
+#include <boost/algorithm/string/predicate.hpp>
```
Comment (MikaelSmith) [inline util/inc/leatherman/util/strings.hpp]: I think you can drop `boost/range/iterator_range.hpp`.

File: windows/inc/leatherman/windows/registry.hpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+         * @param hkey The registry key handle.
+         * @param subkey The name of the registry key.
+         * @param value The name of the registry value.
+         * @return An array of string values corresponding to the REG_MULTI_SZ type.
+         *         Returns an empty vector if the value doesn't exist or isn't a composite string type.
+         */
+        std::vector<std::string> get_registry_strings(HKEY hkey, std::string const& subkey, std::string const& value);
+    }
+
+}}  // leatherman::windows
```
Comment (MikaelSmith) [inline windows/inc/leatherman/windows/registry.hpp]: completely trivial, but the others use `namespace leatherman::windows`.

File: windows/inc/leatherman/windows/wmi.hpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
+/**
+ * @file
+ * Declares utility functions for interacting with Windows Management Instrumentation.
+ */
+#pragma once
+
+#include <leatherman/util/scoped_resource.hpp>
+#include <leatherman/util/strings.hpp>
+#include <boost/algorithm/string/predicate.hpp>
```
Comment (MikaelSmith) [inline windows/inc/leatherman/windows/wmi.hpp]: Both boost/algorithm/string includes aren't needed here.

**Comment by MikaelSmith** [comment]
We should add a README note that the `windows` module requires logging to be setup.

**Comment by MikaelSmith** [comment]
I'd like to eventually have stand-alone tests for these utilities, so filed [LTH-23](https://tickets.puppetlabs.com/browse/LTH-23).

**Pull request merged**